### ISSUE TITLE
Sample code for using same MQTT client from multiple FreeRTOS tasks?

### ISSUE THREAD

**Issue by TepeoIain** [aws/aws-iot-device-sdk-embedded-C#1030]
We are looking for some sample code that shows the correct way to use multiple FreeRTOS tasks to mix device shadow updates with publishes to regular MQTT topics, using the same MQTT client.

Our embedded device uses an Espressif ESP32 module running FreeRTOS v8.2 (for multi-core support) and using the AWS IoT SDK v3.0.1. The firmware uses a number of FreeRTOS tasks to interface with sensors and manage the communications with IoT Core.

One task uses the `aws_iot_shadow_update()` function to periodically perform device shadow updates, containing sensor values and internal device state. Whilst another task uses the `aws_iot_mqtt_publish()` function to publish messages, such as an event alerts, to other topics. The former task is using `aws_iot_shadow_yield()`, and the latter task is using `aws_iot_mqtt_yield()`.

What we find is that even though the period for device shadow updates may be 5 minutes or more (so the MQTT client should be idle for the vast majority of the time) the task sending an alert will sometimes repeatedly receive a `MQTT_CLIENT_NOT_IDLE_ERROR` from `aws_iot_mqtt_publish()`, such that the alert is finally sent several minutes after starting to try to send it. Other times the first time the task tries to send the alert the client will be idle, so the message is published straight away.

Is there a correct way to use multiple FreeRTOS tasks to mix device shadow updates and regular message publishing?

Or is the workaround a single FreeRTOS task that cycles through `aws_iot_shadow_yield()`, `aws_iot_shadow_update()`, `aws_iot_mqtt_yield()` and `aws_iot_mqtt_publish()`?

**Comment by abhidixi11** [comment]
Hello TepeoIain ,
Thanks for contacting us. I'm making sure that I understand your setup  correctly.
1. Based on the description,  are you using CSDK V3 ( master branch ) with FreeRTOS code ?  I'm assuming then you have ported the code from this repo to use with FreeRTOS.  May I know which FreeRTOS version are you using?
2. Are you sharing AWS_IoT_Client between shadow and publish ? In that case you may just able use aws_iot_shadow_yield() as it is a wrapper around aws_iot_mqtt_yield()
3. How many FreeRTOS tasks are you using for activities described above ?
Thanks.

**Comment by TepeoIain** [comment]
Hi. Yes we're using v3.0.1 of the CSDK, and FreeRTOS v8.2. As we're developing with the ESP-IDF, these are the latest stable versions.

On start-up our FreeRTOS firmware establishes a connection to IoT Core using a `AWS_IoT_Client` and `aws_iot_shadow_init()`.

We have a task (1) that is working with GPIO to monitor sensors. When this tasks detects a change in system state, or sensor value, it sets a FreeRTOS event group bit. Another task (2) is in a loop calling `aws_iot_shadow_yield()` and then checking if the event group bit is set. If it is, then task 2 constructs a device shadow update using `aws_iot_shadow_init_json_document()`, `aws_iot_shadow_add_reported()` and `aws_iot_finalize_json_document()` before sending it with `aws_iot_shadow_update()` and resetting the event group bit. Task 2 then goes to sleep for several minutes using `vTaskDelayUntil()`, before restarting the loop.

Another task (3) is monitoring system state and sensor values to detect certain conditions. When the condition is detected a JSON document is constructed and placed in a FreeRTOS message queue. A final task (4) is in a tight loop peeking at the message queue, and if there is a JSON document it calls `aws_iot_mqtt_yield()` and then `aws_iot_mqtt_publish()` using the same `AWS_IoT_Client` as task 2. More often than not this is where the MQTT client reports a `MQTT_CLIENT_NOT IDLE_ERROR`, so the task calls `aws_iot_mqtt_yield()` again until the MQTT client is idle. If the publish is successful then the message is received from the queue, and the task goes back to peeking at the queue for the next message to publish.

Hope that gives some insight into the way we're trying to use multiple FreeRTOS tasks and the same MQTT client to send device shadow updates and publish to regular topics.

**Comment by ravibhagavandas** [comment]
Hi,
From your setup it seems that multiple tasks are trying to compete for the `AWS_IoT_Client` which only supports one operation at a time. One way of doing this is to have a task which waits on a queue for shadow updates or mqtt publish commands. The sensor task and monitor task can post to the queue with the appropriate command types whenever they are ready.
The task can then perform all the required shadow updates and mqtt publishes and then yield for incoming messages for a period of time. As @abhidixi11  pointed you could call `aws_iot_shadow_yield`  to yield for both shadow messages as well as MQTT messages. For the purpose of sending keep alive messages as well as for shorter response time for incoming messages you may need to yield at regular intervals. You may need to adjust the queue wait times accordingly, so that the application always yields at regular intervals even if there are no messages in the queue.

**Comment by TepeoIain** [comment]
Thanks @abhidixi11 and @ravibhagavandas. We're going to amalgamate tasks 2 and 4, so that there's only ever one task trying to access the `AWS_IoT_Client`.

We can't be the only team mixing device shadow updates and regular MQTT topics, so it would be great if there was a template for this approach in the `samples` folder. Kind of a mashup of `subscribe_publish_sample` and `shadow_sample`. Might also be useful to annotate the [architecture diagram](http://aws-iot-device-sdk-embedded-c-docs.s3-website-us-east-1.amazonaws.com/ArchitectureSDK.html), which implies multiple access to the `AWS_IoT_Client`. How, and who to, would I submit those suggestions?

**Comment by abhidixi11** [comment]
Hello @TepeoIain 
Thanks for the suggestion. We are currently focusing on developing updated MQTT library ( in development branch ). We will provide such example for the updated MQTT library.  I have marked this issue as feature request.  
We will be closing this issue if we don't hear back from you,  but please feel free to open another one if you have any further questions. 
Thank you !
